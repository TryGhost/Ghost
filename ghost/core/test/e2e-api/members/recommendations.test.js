const assert = require('assert/strict');
const {agentProvider, mockManager, fixtureManager, matchers, configUtils} = require('../../utils/e2e-framework');
const {anyEtag} = matchers;
const recommendationsService = require('../../../core/server/services/recommendations');
const {Recommendation} = require('@tryghost/recommendations');

async function testClicked({recommendationId, memberId}, test) {
    const before = await recommendationsService.clickEventRepository.getAll({
        filter: `recommendationId:'${recommendationId}'`
    });

    await test();

    const after = await recommendationsService.clickEventRepository.getAll({
        filter: `recommendationId:'${recommendationId}'`
    });

    assert.equal(after.length, before.length + 1);

    // Check member is set
    const added = after.find(event => !before.find(e => e.id === event.id));
    assert.equal(added.memberId, memberId);
}

async function testNotClicked(test) {
    const before = await recommendationsService.clickEventRepository.getCount();
    await test();
    const after = await recommendationsService.clickEventRepository.getCount();

    assert.equal(after, before);
}

async function testNotSubscribed(test) {
    const before = await recommendationsService.subscribeEventRepository.getCount();
    await test();
    const after = await recommendationsService.subscribeEventRepository.getCount();

    assert.equal(after, before);
}

async function testSubscribed({recommendationId, memberId}, test) {
    const before = await recommendationsService.subscribeEventRepository.getAll({
        filter: `recommendationId:'${recommendationId}'`
    });
    await test();
    const after = await recommendationsService.subscribeEventRepository.getAll({
        filter: `recommendationId:'${recommendationId}'`
    });

    assert.equal(after.length, before.length + 1);

    // Check member is set
    const added = after.find(event => !before.find(e => e.id === event.id));
    assert.equal(added.memberId, memberId);
}

describe('Recommendation Event Tracking', function () {
    let membersAgent, membersAgent2, memberId;
    let recommendationId;
    let clock;

    before(async function () {
        membersAgent = await agentProvider.getMembersAPIAgent();
        membersAgent2 = membersAgent.duplicate();
        await membersAgent2.loginAs('authenticationtest@email.com');
        await fixtureManager.init('posts', 'members');

        const membersService = require('../../../core/server/services/members');
        const memberRepository = membersService.api.members;

        const member = await memberRepository.get({email: 'authenticationtest@email.com'});
        memberId = member.id;

        // Add recommendation
        const recommendation = Recommendation.create({
            title: `Recommendation`,
            description: `Description`,
            url: new URL(`https://recommendation.com`),
            favicon: null,
            featuredImage: null,
            excerpt: null,
            oneClickSubscribe: false
        });

        await recommendationsService.repository.save(recommendation);
        recommendationId = recommendation.id;
    });

    beforeEach(function () {
        mockManager.mockMail();
    });

    afterEach(async function () {
        clock?.restore();
        clock = undefined;
        await configUtils.restore();
        mockManager.restore();
    });

    describe('Authenticated', function () {
        it('Can track subscribe clicks', async function () {
            await testNotClicked(async () => {
                await testSubscribed({recommendationId, memberId}, async () => {
                    await membersAgent2
                        .post('/api/recommendations/' + recommendationId + '/subscribed/')
                        .body({})
                        .expectStatus(204)
                        .matchHeaderSnapshot({
                            etag: anyEtag
                        })
                        .expectEmptyBody();
                });
            });
        });

        it('Can track clicks', async function () {
            await testNotSubscribed(async () => {
                await testClicked({recommendationId, memberId}, async () => {
                    await membersAgent2
                        .post('/api/recommendations/' + recommendationId + '/clicked/')
                        .body({})
                        .expectStatus(204)
                        .matchHeaderSnapshot({
                            etag: anyEtag
                        })
                        .expectEmptyBody();
                });
            });
        });
    });

    describe('Unauthenticated', function () {
        it('Can not track subscribe clicks', async function () {
            await testNotClicked(async () => {
                await testNotSubscribed(async () => {
                    await membersAgent
                        .post('/api/recommendations/' + recommendationId + '/subscribed/')
                        .body({})
                        .expectStatus(401)
                        .matchHeaderSnapshot({
                            etag: anyEtag
                        });
                });
            });
        });

        it('Can track clicks', async function () {
            await testNotSubscribed(async () => {
                await testClicked({recommendationId, memberId: null}, async () => {
                    await membersAgent
                        .post('/api/recommendations/' + recommendationId + '/clicked/')
                        .body({})
                        .expectStatus(204)
                        .matchHeaderSnapshot({
                            etag: anyEtag
                        })
                        .expectEmptyBody();
                });
            });
        });
    });
});
