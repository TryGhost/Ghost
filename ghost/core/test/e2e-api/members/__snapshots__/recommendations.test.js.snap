// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendation Event Tracking Authenticated Can track clicks 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "x-powered-by": "Express",
}
`;

exports[`Recommendation Event Tracking Authenticated Can track subscribe clicks 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "x-powered-by": "Express",
}
`;

exports[`Recommendation Event Tracking Authenticated Cannot track invalid types 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "226",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Recommendation Event Tracking Unauthenticated Can not track subscribe clicks 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "206",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Recommendation Event Tracking Unauthenticated Can track clicks 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "x-powered-by": "Express",
}
`;

exports[`Recommendation Event Tracking Unauthenticated Cannot track invalid types 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "226",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;
